plugin/trino-hive/src/main/java/io/trino/plugin/hive/ion/IonFileWriter.java

```diff
@@ -0,0 +1,130 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.hive.ion;
+
+import com.amazon.ion.IonWriter;
+import com.google.common.io.CountingOutputStream;
+import io.trino.hive.formats.compression.CompressionKind;
+import io.trino.hive.formats.ion.IonEncoder;
+import io.trino.hive.formats.ion.IonEncoderFactory;
+import io.trino.hive.formats.line.Column;
+import io.trino.memory.context.AggregatedMemoryContext;
+import io.trino.plugin.hive.FileWriter;
+import io.trino.spi.Page;
+import io.trino.spi.TrinoException;
+import io.trino.spi.type.TypeManager;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.List;
+import java.util.Optional;
+import java.util.function.LongSupplier;
+
+import static io.trino.plugin.hive.HiveErrorCode.HIVE_WRITER_CLOSE_ERROR;
+import static io.trino.plugin.hive.HiveErrorCode.HIVE_WRITER_DATA_ERROR;
+
+public class IonFileWriter
+        implements FileWriter
+{
+    private final AggregatedMemoryContext outputStreamMemoryContext;
+    private final Closeable rollbackAction;
+    private final IonEncoder pageEncoder;
+    private final IonWriter writer;
+    private final OutputStream outputStream;
+    private final LongSupplier bytesWritten;
+
+    public IonFileWriter(
```

Comment:
Check for nulls

Reply:
I added checks where I believed they might be needed, also on the PageSource side as well.
I'm assuming the concern is around the nullness carrying over into the stateful object and becoming an error later in program flow; null references which would raise in the same method are not a problem.
I think I got it right given that.